Generalize parameter verification and broadcast

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75977
* #75976
* __->__ #75970
* #75945

Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

rohan-varma committed:
Generalize parameter verification and broadcast

Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

awgu commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,31 @@
+def _verify_param_shape_across_processes(process_group, tensors, logger=None):
+    return dist._verify_params_across_processes(process_group, tensors, logger)
+
+def _sync_params_and_buffers(
```
nit: Do we want to add some type annotations?

rohan-varma replied:
I think we should probably decide whether we want to invest in typing across distributed codebase and take that on as a BE task?

awgu commented on torch/distributed/utils.py:
```diff
@@ -0,0 +1,31 @@
+import torch.distributed as dist
+
+def _verify_param_shape_across_processes(process_group, tensors, logger=None):
```
nit: `_verify_param_shapes_across_processes()` since verifying multiple parameters shapes?

awgu reviewed (approved):
Looks good to me! Just some nits.

zhaojuanmao reviewed (approved):
nice generalization!

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

zhaojuanmao commented:
by the way, let's do benchmarks before landing as well just in case

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Generalize parameter verification and broadcast"


Generalizes the two verifcation / module sync APIs out of DDP to be
consumed by FSDP.

Differential Revision: [D35712025](https://our.internmc.facebook.com/intern/diff/D35712025/)

[ghstack-poisoned]
